File: src/n10_slam/scripts/odometry_node.py
#!/usr/bin/env python3
import rospy
import math
from nav_msgs.msg import Odometry
from geometry_msgs.msg import Twist, TransformStamped
from tf2_msgs.msg import TFMessage

class OdomNode:
    def __init__(self):
        rospy.init_node('odometry_node')
        
        self.x = 0.0
        self.y = 0.0
        self.th = 0.0
        self.vx = 0.0
        self.vth = 0.0
        self.last_time = rospy.Time.now()
        
        self.tf_pub = rospy.Publisher('/tf', TFMessage, queue_size=50)
        self.odom_pub = rospy.Publisher('/odom', Odometry, queue_size=50)
        rospy.Subscriber('/sensor/cmd_vel', Twist, self.vel_cb)
        
        rospy.loginfo("里程计启动 - 高频率发布")

    def vel_cb(self, msg):
        self.vx = -msg.linear.x
        self.vth = -msg.angular.z

    def update(self):
        now = rospy.Time.now()
        dt = (now - self.last_time).to_sec()
        if dt <= 0 or dt > 0.5:
            self.last_time = now
            return
        
        self.x += self.vx * math.cos(self.th) * dt
        self.y += self.vx * math.sin(self.th) * dt
        self.th += self.vth * dt
        
        if self.th > math.pi: self.th -= 2*math.pi
        if self.th < -math.pi: self.th += 2*math.pi
        
        self.last_time = now
        
        qz = math.sin(self.th/2)
        qw = math.cos(self.th/2)
        
        # 发布TF
        t = TransformStamped()
        t.header.stamp = now
        t.header.frame_id = 'odom'
        t.child_frame_id = 'base_link'
        t.transform.translation.x = self.x
        t.transform.translation.y = self.y
        t.transform.translation.z = 0.0
        t.transform.rotation.x = 0.0
        t.transform.rotation.y = 0.0
        t.transform.rotation.z = qz
        t.transform.rotation.w = qw
        
        tf_msg = TFMessage()
        tf_msg.transforms.append(t)
        self.tf_pub.publish(tf_msg)
        
        # 发布Odom
        odom = Odometry()
        odom.header.stamp = now
        odom.header.frame_id = 'odom'
        odom.child_frame_id = 'base_link'
        odom.pose.pose.position.x = self.x
        odom.pose.pose.position.y = self.y
        odom.pose.pose.position.z = 0.0
        odom.pose.pose.orientation.x = 0.0
        odom.pose.pose.orientation.y = 0.0
        odom.pose.pose.orientation.z = qz
        odom.pose.pose.orientation.w = qw
        odom.twist.twist.linear.x = self.vx
        odom.twist.twist.angular.z = self.vth
        self.odom_pub.publish(odom)
        
        rospy.loginfo_throttle(1, "Odom: x=%.2f y=%.2f th=%.0f vx=%.2f vth=%.2f", 
                              self.x, self.y, math.degrees(self.th), self.vx, self.vth)

    def run(self):
        # 提高到100Hz
        rate = rospy.Rate(100)
        while not rospy.is_shutdown():
            self.update()
            rate.sleep()

if __name__ == '__main__':
    try:
        OdomNode().run()
    except rospy.ROSInterruptException:
        pass

File: src/n10_slam/src/odometry_node.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <geometry_msgs/Twist.h>
#include <tf2_ros/transform_broadcaster.h>
#include <geometry_msgs/TransformStamped.h>
#include <cmath>

class OdomNode {
public:
    OdomNode() : x_(0.0), y_(0.0), th_(0.0), vx_(0.0), vth_(0.0) {
        ros::NodeHandle nh;
        odom_pub_ = nh.advertise<nav_msgs::Odometry>("/odom", 50);
        vel_sub_  = nh.subscribe("/sensor/cmd_vel", 10, &OdomNode::velCb, this);
        last_time_ = ros::Time::now();
        ROS_INFO("里程计节点启动 (C++) - 100Hz");
    }

    void velCb(const geometry_msgs::Twist::ConstPtr& msg) {
        vx_  = -msg->linear.x;
        vth_ = -msg->angular.z;
    }

    void update() {
        ros::Time now = ros::Time::now();
        double dt = (now - last_time_).toSec();
        if (dt <= 0.0 || dt > 0.5) {
            last_time_ = now;
            return;
        }

        x_  += vx_ * std::cos(th_) * dt;
        y_  += vx_ * std::sin(th_) * dt;
        th_ += vth_ * dt;

        // 归一化到 [-π, π]
        while (th_ >  M_PI) th_ -= 2.0 * M_PI;
        while (th_ < -M_PI) th_ += 2.0 * M_PI;

        last_time_ = now;

        double qz = std::sin(th_ / 2.0);
        double qw = std::cos(th_ / 2.0);

        // 发布 TF
        geometry_msgs::TransformStamped tf;
        tf.header.stamp    = now;
        tf.header.frame_id = "odom";
        tf.child_frame_id  = "base_link";
        tf.transform.translation.x = x_;
        tf.transform.translation.y = y_;
        tf.transform.translation.z = 0.0;
        tf.transform.rotation.x = 0.0;
        tf.transform.rotation.y = 0.0;
        tf.transform.rotation.z = qz;
        tf.transform.rotation.w = qw;
        tf_broadcaster_.sendTransform(tf);

        // 发布 Odometry
        nav_msgs::Odometry odom;
        odom.header.stamp    = now;
        odom.header.frame_id = "odom";
        odom.child_frame_id  = "base_link";
        odom.pose.pose.position.x  = x_;
        odom.pose.pose.position.y  = y_;
        odom.pose.pose.position.z  = 0.0;
        odom.pose.pose.orientation.x = 0.0;
        odom.pose.pose.orientation.y = 0.0;
        odom.pose.pose.orientation.z = qz;
        odom.pose.pose.orientation.w = qw;
        odom.twist.twist.linear.x  = vx_;
        odom.twist.twist.angular.z = vth_;
        odom_pub_.publish(odom);

        ROS_INFO_THROTTLE(1.0, "Odom: x=%.2f y=%.2f th=%.0f° vx=%.2f vth=%.2f",
                          x_, y_, th_ * 180.0 / M_PI, vx_, vth_);
    }

    void run() {
        ros::Rate rate(100);
        while (ros::ok()) {
            ros::spinOnce();
            update();
            rate.sleep();
        }
    }

private:
    ros::Publisher  odom_pub_;
    ros::Subscriber vel_sub_;
    tf2_ros::TransformBroadcaster tf_broadcaster_;

    double x_, y_, th_;
    double vx_, vth_;
    ros::Time last_time_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "odometry_node");
    OdomNode node;
    node.run();
    return 0;
}
